src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs

```diff
@@ -19,24 +19,31 @@ internal sealed class SafeFreeCertContext : DebugSafeHandle
     internal sealed class SafeFreeCertContext : SafeHandle
     {
 #endif
-        public SafeFreeCertContext(IntPtr handle) : base(handle, true)
+        private readonly SafeX509Handle _certificate;
+
+        public SafeFreeCertContext(SafeX509Handle certificate) : base(IntPtr.Zero, true)
         {
+            if ((null != certificate) && !certificate.IsInvalid)
+            {
+                _certificate = Interop.libcrypto.X509_dup(certificate);
```

Comment:
Actually, no, since it's a SafeHandle returned now it's... probably fine.

* SSL_get_peer_certificate bumps the refcount to 2, that is encapsulated in the returned SafeHandle.
* This code calls X509_dup, bumping the refcount to 3.
* The originally returned SafeHandle gets Finalized (since you're not Disposing it), decrementing the refcount to 2.
* The output of X509_dup is eventually Disposed (or FInalized), decrementing the refcount to 1.
* The SafeSslHandle is eventually Disposed (or Finalized), decrementing the refcount to 0 and freeing the native resource.

So really the problem is that you're sending a SafeHandle off to the finalizer queue on every call.

The point about needing to know what OpenSSL does still stands, though; if the p/invoke target return type was IntPtr instead of SafeHandle there'd be a decrement missing from that series.